### PR TITLE
split test suite

### LINKED ISSUE #427: split the test-suite
## situation 

having a test-suite is fine, but it has a flaws: it is a mix of general rules for the purl spec, and the specific rules for the fast-living and still-evolving type-spec.

## problems 

- this current situation makes it hard to use the test-suite for general-purpose libraries that dont implement every purl-type specific rule, but only the core spec.  
  For example, the PHP implementation does not have all these purl-type specific rules implemented, it expects the downstream user to take care of the spec they use, since the downstream user knows their domain.
- this current situation makes it hard to update their version of the test-suite for general-purpose libraries that want to use the latest test suite without adding the new purl-types ("features").  
  For example, the Java implementation can not simply pull the latest test suite that includes the latest general purl-spec related cases, since the test-suite also includes cases for purl-specifics that are not yet implemented in Java library.

## proposal

split the test suite into core spec, and purl-type specifics,
so that implementations can pick(glob) the relevant parts of the test suite they actually implement.

## suggestion

- have a folder, that contains all test suites.
- have a file for the purl core spec: `PURL-SPECIFICATION.json`
- have a file for each purl type: `PURL-TYPES.<type>.json`

```text
test-suite
    ├── PURL-SPECIFICATION.json
    ├── PURL-TYPES.apk.json
    ├── PURL-TYPES.bitnami.json
...
    ├── PURL-TYPES.generic.json
...
    └── PURL-TYPES.swift.json

```

### PR DESCRIPTION
**Pull request by jkowalleck**
- have a folder, that contains all test suites.
- have a file for the purl core spec: `PURL-SPECIFICATION.json`
- have a file for each purl type: `PURL-TYPES.<type>.json`

```text
test-suite-data
    ├── PURL-SPECIFICATION.json
    ├── PURL-TYPES.apk.json
    ├── PURL-TYPES.bitnami.json
...
    ├── PURL-TYPES.generic.json
...
    └── PURL-TYPES.swift.json
```

----

> [!NOTE]
> this PR does not add new or alter existing tests. it just refactors their location

----

fixes #427

TODO / DONE

- [x] split test suite per type via automatism/script
- [x] extract ALL the general cases manually afterward

created via script:
```py
#!/usr/bin/env python3

from json import loads as json_loads, dumps as json_dumps
from urllib.request import urlopen
from os.path import join
from typing import Any, List, Dict

TestCaseType = Dict[str, Any]

mixed_test_suite_url = 'https://raw.githubusercontent.com/package-url/purl-spec/refs/heads/main/test-suite-data.json'
with urlopen(mixed_test_suite_url) as mixed_test_suite_res:
    mixed_test_suite_data: List[TestCaseType] = json_loads(mixed_test_suite_res.read())

ts_types: Dict[str, List[TestCaseType]] = {'generic': []}

for tc in mixed_test_suite_data:
    if tc['type'] is None:
        print('none_case:', json_dumps(tc, indent=2), '', sep='\n')
        continue
    tc_type = str(tc['type']).lower()
    if tc_type not in ts_types:
        ts_types[tc_type] = [tc]
    else:
        ts_types[tc_type].append(tc)

for tct, tcs in ts_types.items():
    with open(join('test-suite', f'PURL-TYPES.{tct}.json'), 'wt') as tcf:
        tcf.write(json_dumps(tcs, indent=2))
```

### CODE REVIEW

File: test-suite/PURL-SPECIFICATION.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+[
+
```
Comment (jkowalleck) [inline test-suite/PURL-SPECIFICATION.json:2]: TODO

Comment (jkowalleck) [reply test-suite/PURL-SPECIFICATION.json]: done

File: test-suite-data/PURL-TYPES.maven.json [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    "purl": "pkg://maven/org.apache.commons/io",
+    "canonical_purl": "pkg:maven/org.apache.commons/io",
+    "type": "maven",
+    "namespace": "org.apache.commons",
+    "name": "io",
+    "version": null,
+    "qualifiers": null,
+    "subpath": null,
+    "is_invalid": false
+  },
```
Comment (matt-phylum) [inline test-suite-data/PURL-TYPES.maven.json:94]: Some of these tests are duplicates of core tests and aren't related to the package type.

**Review by matt-phylum — COMMENTED** [review]
The core file is missing tests for canonical encoding of all characters in all the different positions. This isn't currently defined by the spec for all characters, but at least there should be tests for the characters that are defined and for some critical cases where failure to encode and decode properly results in unexpected errors or incorrect values
- plus vs space (broken for at least 4 implementations)
- percent signs (broken for 1 implementation)
- ampersands in qualifier values (broken for 2 implementations)
- a control character < 0x10 (broken for 2 implementations)
- a BMP Unicode character (< U+10000, eg `☃`, broken for 2 implementations)
- a non-BMP Unicode character (>= U+10000, eg `💩`, broken for 2 implementations).

Some of encoding cases used to be covered and are not anymore. For example, colons are now only tested for implementations that support pkg:docker and pkg:cpan. Slashes in qualifier values are only tested for implementations that support huggingface or mlflow.

**Review by mprpic — APPROVED** [review]
LGTM, I would file separate issues for the comments that @matt-phylum left in this PR since the change here is only about better organization of the test data.

**Comment by jkowalleck** [comment]
@pombredanne @package-url/purl-spec-helpers can we merge this one?

I mean, we could wait until all the tens open PRs that want to modify the test suite are merged eventually, but this means we would not have a reasonably usable test suite.
Or wait until we have clarification for certain edge cases in the test suite, whichseams to take months.

Anyway, I am a huge fan of many small, scoped changes - like this one - so downstream users can review and adopt in small iterations, instead of one big blob of rework.

Let's merge this one, so that downstream users have an immediate benefit, while all these other PRs are still in triage months.

**Review by mprpic — DISMISSED** [review]

**Comment by jkowalleck** [comment]
need to incorporate https://github.com/package-url/purl-spec/pull/416

File: test-suite-data/PURL-TYPES.maven.json [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+  {
+    "description": "maven often uses qualifiers",
+    "purl": "pkg:Maven/org.apache.xmlgraphics/batik-anim@1.9.1?classifier=sources&repositorY_url=repo.spring.io/release",
```
Comment (ppkarwasz) [inline test-suite-data/PURL-TYPES.maven.json:28]: Isn't this an **invalid** example, since the solidus `/` in `repo.spring.io/release` is not percent encoded?

Comment (matt-phylum) [reply test-suite-data/PURL-TYPES.maven.json:28]: No. The parser is supposed to split on `?` before `/`.

**Review by johnmhoran — APPROVED** [review]
Excellent idea @jkowalleck -- LGTM.

File: test-suite-data/PURL-SPECIFICATION.json [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    "version": "2.3.0",
+    "qualifiers": null,
+    "subpath": null,
+    "is_invalid": false
+  },
+  {
+    "description": "invalid encoded colon : between scheme and type",
+    "purl": "pkg%3Amaven/org.apache.commons/io",
+    "canonical_purl": null,
+    "type": "maven",
```
[the remark this answers is not in the document]
Comment (johnmhoran) [reply test-suite-data/PURL-SPECIFICATION.json:126]: @matt-phylum Thanks for your comments on this test, which seems to be a variation of the test I added as part of my (merged) `scheme` PR https://github.com/package-url/purl-spec/pull/361.  I had questions at the time re how to properly construct the input values when the PURL is meant to be invalid, and similar questions re the 3 is-invalid-true tests in my (merged) `type` PR https://github.com/package-url/purl-spec/pull/383.  Subsequent discussion led to 
- @jkowalleck's https://github.com/package-url/purl-spec/issues/403 and 
- my related https://github.com/package-url/purl-spec/issues/419.

Clarifying the core spec's [test-construction instructions](https://github.com/package-url/purl-spec/blob/main/PURL-SPECIFICATION.rst#tests) would be a valuable next step in the test-updating process.  (I suspect the test-construction clarification could be done even while additional work on the "Character encoding" section continues.)

File: test-suite-data/PURL-TYPES.maven.json (hunk 3 above)
Comment (ppkarwasz) [reply test-suite-data/PURL-TYPES.maven.json:28]: What about #439? I assume that `/` must be unencoded, when used as a PURL separator and encoded otherwise?

**Comment by jkowalleck** [comment]
superseded by #514
